Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -173,6 +173,108 @@ public struct PSKClientIdentityResponse: Sendable {
+/// host name indicated in the TLS client hello, while the provided `NIOSSLContext` will be the current
+/// server context being used for the handshake.
+///
+/// Within this callback, the user can create and return a new server `NIOSSLContext` for the given host,
+/// return the same context provided if it is sufficient to complete the handshake, or return `throw` in the event
+/// of an exception.
+///
+public typealias NIOSSLContextCallback = @Sendable (NIOSSLClientExtensionValues, EventLoopPromise<NIOSSLContext>) -> Void
+
+/// A struct that provides helpers for working with a NIOSSLContextCallback.
+internal struct CustomContextManager {
+    private let callback: NIOSSLContextCallback
+
+    private var state: State
+
+    init(callback: @escaping NIOSSLContextCallback) {
+        self.callback = callback
+        self.state = .notStarted
+    }
+}
+
+extension CustomContextManager {
+    private enum State {
+        case notStarted
+
+        case pendingResult
+
+        case complete(Result<NIOSSLContext, Error>)
+    }
+}
+
+extension CustomContextManager {
+    mutating func loadContext(ssl: OpaquePointer) -> Result<NIOSSLContext, Error>? {
+        switch state {
+        case .pendingResult:
+            // In the pending case we just return nil
+            return nil
+        case .complete(let result):
+            // In the complete we can return our result
+            return result
+        case .notStarted:
+            // Load the attached connection so we can resume handshake when future resolves
+            let connection = SSLConnection.loadConnectionFromSSL(ssl)
+
+            guard let eventLoop = connection.eventLoop else {
+                preconditionFailure("""
+                    SSL_CTX_set_cert_cb was executed without an event loop assigned to the connection.
+                    This should not be possible, please file an issue.
+                """)
+            }
+            
+            // Construct extension values to be passed to callback
+            let cServerHostname = CNIOBoringSSL_SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name)
+            let serverHostname = cServerHostname.map { String(cString: $0) }
+            let values = NIOSSLClientExtensionValues(serverHostname: serverHostname)
+
+            // We're responsible for creating the promise and the user provided callback will fulfill it
+            let promise = eventLoop.makePromise(of: NIOSSLContext.self)
+            self.callback(values, promise)
+
+            // After invoking the user callback we can update our state to pending
+            self.state = .pendingResult
+
+            // Ensure we execute any completion on the next event loop tick
+            // This ensures that we suspend before calling resume
+            eventLoop.execute {
+                promise.futureResult.whenComplete { result in
```

Comment:
These two calls are inverted: we need to `execute` inside the `whenComplete`, not the other way around.